lib/filewatch/since_db.rb

```diff
@@ -0,0 +1,247 @@
+require 'filewatch/boot_setup' unless defined?(FileWatch)
+
+module FileWatch
+  class SinceDb
+    def self.parse_line(line)
+      parts = line.split(" ")
+      parse_line_v2(parts) || parse_line_v1(parts)
+    end
+
+    def self.parse_line_v2(parts)
+      return false if !parts.first.include?(",")
+      fp1 = parts.shift.split(",").map(&:to_i)
+      fp1[0] = Fnv.coerce_bignum(fp1[0])
+      pos = parts.shift.to_i
+      exp = Float(parts.shift)
+      if !parts.empty?
+        fp2 = parts.shift.split(",").map(&:to_i)
+      end
+      [SincedbKey2.new(*fp1), SincedbValue.new(pos, exp, fp2)]
+    end
+
+    def self.parse_line_v1(parts)
+      kparts = parts.shift(3).map(&:to_i)
+      pos, exp = parts
+      exp = Float(exp) if !exp.nil?
+      [SincedbKey1.new(*kparts), SincedbValue.new(pos.to_i, exp)]
+    end
+
+    def self.serialize_kv(k, v)
+      "#{k} #{v}"
+    end
+
+    attr_reader :path, :short_keys
+    attr_writer :logger
+
+    def initialize(opts, loggr)
+      @logger = loggr
+      @path = opts[:sincedb_path]
+      @lastwarn = Hash.new { |h, k| h[k] = 0 }
+      @sincedb_last_write = 0
+      @interval = opts[:sincedb_write_interval]
+      @sincedb = {}
+      @short_keys = Hash.new{|h, k| h[k] = []}
+    end
+
+    def write_periodically
+      now = Time.now.to_i
+      delta = now - @sincedb_last_write
+      if delta >= @interval
+        @logger.debug? && @logger.debug("writing sincedb (delta since last write = #{delta})")
+        sincedb_write
+        @sincedb_last_write = now
+      end
+    end
+
+    def write(reason=nil)
+      @logger.debug? && @logger.debug("caller requested sincedb write (#{reason})")
+      sincedb_write
+    end
+
+    def open
+      @time_sdb_opened = Time.now.to_f
+      begin
+        File.open(path) do |db|
+          @logger.debug? && @logger.debug("SinceDb open: reading from #{path}")
+          db.each {|line| store_from_file(line) }
+        end
+
+        @logger.debug? && @logger.debug("SinceDb open: keys read #{@sincedb.keys.inspect}")
+      rescue => e
+        #No existing sincedb to load
+        @logger.debug? && @logger.debug("SinceDb open: error: #{path}: #{e.inspect}")
+      end
+    end
+
+    def deallocate(wf)
+      return if !member?(wf.storage_key)
+      get(wf.storage_key).deallocate_watched_file
+    end
+
+    def find(wf, event)
```

Comment:
Do we really need to pass an `event` to the find method? The only use case I see is for a debug statement?